### ISSUE TITLE
[Bug]: `Navigate` causes an infinite loop if it remains mounted after navigation

### ISSUE THREAD

**Issue by jordan-burnett** [remix-run/react-router#8733]
### What version of React Router are you using?

6.2.1

### Steps to Reproduce

In a previous version of React router we could test components that used `Redirect` in isolation without needing to mock the implementation details of react-router like below:

```javascript
const ErrorRedirect = ({ error }) =>  {
    if(error) {
       return <Redirect to="/error" />
   }
   return null;
}

it('redirects to the error route if it receives an error', () => {
    const history = createMemoryHistory();
    render(
        <HistoryRouter history={history}>
            <ErrorRedirect error={new Error()} />
         </HistoryRouter>
    );
    expect(history.location.pathname).toBe('/error');
});

```

When switching `Redirect` for a `Navigate` component in v6 this same test causes the app to crash.  This is because `Navigate` causes an infinite loop if it remains mounted after the navigation has taken place. 

### Expected Behavior

Would expect the location to change but the component to only rerender once.

### Actual Behavior

`Navigate` causes an infinite loop and the process crashes.

**Comment by rajis31** [comment]
Yes experiencing the same issue when trying to create a Private Route component to check if the user is authenticated or not before going to the destination. Adding the **Navigate** component is causing my app to render in an infinite loop.

**Comment by codercodingthecode** [comment]
I had a similar issue, Add the prop ```replace```, which seems to be fixed the issue for me.
```javascript
     ...
     
      <Route path="tosomething" element={<Navigate to={'/someroute/' + userid} replace />} />
      
    ...

**Comment by jordan-burnett** [comment]
As a workaround we've implemented the fix in https://github.com/remix-run/react-router/pull/8734 as a standalone component.  You should be able to replace the built-in version of `Navigate` with the following until a fix is merged:

```typescript
import type { NavigateProps } from 'react-router-dom';
import { useNavigate } from 'react-router-dom';

export const Navigate: React.FC<NavigateProps> = ({ to, replace, state }) => {
  const navigate = useNavigate();

  useEffect(() => {
    navigate(to, { replace, state });
  }, [navigate, to, replace, state]);

  return null;
};
```

**Comment by monzilnepali** [comment]
@jordan-burnett Looks great. Thanks

**Comment by cborchert** [comment]
Any update on this ? We just realized that a recent slowdown on one of our products is due to this error. I plan on using @jordan-burnett's workaround for the moment, but we would love for react-router-dom to patch this as soon as possible as it does have tangible effects on in-prod code.

**Comment by ryanflorence** [comment]
This doesn't seem broken to me. The condition that causes the redirect should be resolved. It's like saying that `while(true)` shouldn't cause an infinite loop isn't it?

**Comment by cborchert** [comment]
@ryanflorence  the current code for the Navigate component does not have a dependency array, which means that `navigate` is called on every render of the parent component

https://github.com/remix-run/react-router/blob/f3d87dcc91fbd6fd646064b88b4be52c15114603/packages/react-router/lib/components.tsx#L93-L95

While I guess you could argue that the responsibility is on the user to guarantee that the parent component is not re-rendered 
 too often, or as a result of this use effect, I think you could also argue that when we use the `<Navigate />` component, the expected behavior is a single navigation event (which was the previous behavior).

**Comment by ryanflorence** [comment]
It's up to the app to fix its own infinite redirect logic, otherwise a valid case where a redirect ends up in a second redirect of the same component but a different case will break.

```jsx
if (foo || bar) {
  return <Navigate to="/somewhere" />
}
```

If `foo` causes a redirect, is resolved on the way back around the second time, and then `bar` is true, then `Navigate` won't work.

**Comment by jordan-burnett** [comment]
> It's up to the app to fix its own infinite redirect logic, otherwise a valid case where a redirect ends up in a second redirect of the same component but a different case will break.
> 
> ```js
> if (foo || bar) {
>   return <Navigate to="/somewhere" />
> }
> ```
> 
> If `foo` causes a redirect, is resolved on the way back around the second time, and then `bar` is true, then `Navigate` won't work.

@ryanflorence  Sorry for the late response.  What would break here? My understanding is that with a dependency array added, `Navigate` will only fire once if the (foo || bar) condition resolves unless the  `to` prop changes, which to me would be the expected behaviour.

Appreciate that may not have been how the new `Navigate` was intended to work though.  My expectation was it would behave the same as the previous `Redirect` component, which isn't the case, but we can work around it with our own wrapper for now.

**Comment by vicodinvic1** [comment]
> As a workaround we've implemented the fix in #8734 as a standalone component. You should be able to replace the built-in `Navigate` component with something like the below, or just use `useNavigate` directly in your components until then.
> 
> ```ts
> import type { NavigateProps } from 'react-router-dom';
> import { useNavigate } from 'react-router-dom';
> 
> export const Navigate: React.FC<NavigateProps> = ({ to, replace, state }) => {
>   const navigate = useNavigate();
>   const navigateRef = useRef(navigate);
> 
>   useEffect(() => {
>     navigateRef.current = navigate;
>   }, [navigate]);
> 
>   useEffect(() => {
>     navigateRef.current(to, { replace, state });
>   }, [to, replace, state]);
> 
>   return null;
> };
> ```

this code works, but it causes page reload on each navigate call
how to handle this issue?

tested on 6.4.0 react-router version

**Comment by shamseerahammedm** [comment]
How do u get this working with `createBrowserRouter` ? its causing infinite loop. Tried solution by @jordan-burnett , but then it fixes inifinte loop but nothing is rendering on page, just shows a blank page 

```
const getRoutes = (isUserLoggedIn : boolean) => {
  console.log('isUserLoggedIn',isUserLoggedIn);
  const router = createBrowserRouter([
    {
      path: '/',
      element: isUserLoggedIn ? <Navigate to="/dashboard" replace={true} /> : <LoginPage/>,
      children: [
        {
          path: '/dashboard',
          element: (
            <DashboardPage/>
          ),
        },
      ],
    },
  ]);
  return router;
};
```

```
 <RouterProvider router={getRoutes(isUserLoggedIn)} />
 ```
